// es-toolkit/dist/math/index.js
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const require_random = require("./random.js");
const require_randomInt = require("./randomInt.js");
const require_clamp = require("./clamp.js");
const require_inRange = require("./inRange.js");
const require_sum = require("./sum.js");
const require_mean = require("./mean.js");
const require_sumBy = require("./sumBy.js");
const require_meanBy = require("./meanBy.js");
const require_median = require("./median.js");
const require_medianBy = require("./medianBy.js");
const require_percentile = require("./percentile.js");
const require_range = require("./range.js");
const require_rangeRight = require("./rangeRight.js");
const require_round = require("./round.js");
exports.clamp = require_clamp.clamp;
exports.inRange = require_inRange.inRange;
exports.mean = require_mean.mean;
exports.meanBy = require_meanBy.meanBy;
exports.median = require_median.median;
exports.medianBy = require_medianBy.medianBy;
exports.percentile = require_percentile.percentile;
exports.random = require_random.random;
exports.randomInt = require_randomInt.randomInt;
exports.range = require_range.range;
exports.rangeRight = require_rangeRight.rangeRight;
exports.round = require_round.round;
exports.sum = require_sum.sum;
exports.sumBy = require_sumBy.sumBy;
